contracts/hydro/src/utils.rs

```diff
@@ -0,0 +1,290 @@
+
+    // If we got through here, it means that constants.current_users_extra_cap > 0.
+    // If constants.current_users_extra_cap was 0 and public_cap (equal to total_cap)
+    // is equal to total_locked_tokens, then we would error out at the start of this
+    // function, because any amount_to_lock would exceed the total_cap. This is just
+    // a safety check to make the code resilient to future changes.
+    if constants.current_users_extra_cap == 0 {
+        return lock_limit_reached_error;
+    }
+
+    // If there is still room in extra_cap, then check if this
+    // is a user that should be allowed to use the extra_cap.
+    if !can_user_lock_in_extra_cap(deps, constants, current_round, sender, amount_to_lock)? {
+        return lock_limit_reached_error;
+    }
+
+    Ok(LockingInfo {
+        lock_in_public_cap: None,
+        lock_in_known_users_cap: Some(amount_to_lock),
+    })
+}
+
+fn can_user_lock_in_extra_cap(
+    deps: &DepsMut<NeutronQuery>,
+    constants: &Constants,
+    current_round: u64,
+    sender: &Addr,
+    amount_to_lock: u128,
+) -> Result<bool, ContractError> {
+    let extra_locked_tokens_round_total = EXTRA_LOCKED_TOKENS_ROUND_TOTAL
+        .may_load(deps.storage, current_round)?
+        .unwrap_or(0);
+
+    if extra_locked_tokens_round_total + amount_to_lock > constants.current_users_extra_cap {
+        return Ok(false);
+    }
+
+    // Determine if user has the right to lock in extra_cap by looking at its voting power in previous round.
+    // If we are in round 0 then check the current round.
+    let round_to_check = match current_round {
+        0 => current_round,
+        _ => current_round - 1,
+    };
+
+    // Calculate user's voting power share in the total voting power
+    let users_voting_power = Decimal::from_ratio(
+        get_user_voting_power_for_past_round(
+            &deps.as_ref(),
+            constants,
+            sender.clone(),
+            round_to_check,
+        )?,
+        Uint128::one(),
+    );
+    let total_voting_power = get_total_power_for_round(deps.as_ref(), round_to_check)?;
+
+    // Prevent division by zero or break early in case user had no voting power in previous round.
+    if total_voting_power == Decimal::zero() || users_voting_power == Decimal::zero() {
+        return Ok(false);
+    }
+
+    let users_voting_power_share = users_voting_power.checked_div(total_voting_power)?;
+
+    // Calculate what would be users share of extra locked tokens in the maximum allowed extra locked tokens
+    let users_current_extra_lock = EXTRA_LOCKED_TOKENS_CURRENT_USERS
+        .may_load(deps.storage, (current_round, sender.clone()))?
+        .unwrap_or(0);
+
+    let users_extra_lock =
+        Decimal::from_ratio(users_current_extra_lock + amount_to_lock, Uint128::one());
+    let maximum_extra_lock = Decimal::from_ratio(constants.current_users_extra_cap, Uint128::one());
+    let users_extra_lock_share = users_extra_lock.checked_div(maximum_extra_lock)?;
```

Comment:
this should also have a check around maximum_extra_lock to avoid division by zero, maybe? I don't think this can happen in the current code, but I feel it would benefit from being made safe from division by zero in case we change the code